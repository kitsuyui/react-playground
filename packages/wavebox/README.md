# @kitsuyui/react-wavebox

[![npm version](https://badge.fury.io/js/@kitsuyui%2Freact-wavebox.svg)](https://badge.fury.io/js/@kitsuyui%2Freact-wavebox)

## Demo

Storybook: https://react-playground.storybook.kitsuyui.com/

## Installation

### npm

```sh
npm install @kitsuyui/react-wavebox
```

### yarn

```sh
yarn add @kitsuyui/react-wavebox
```

### pnpm

```sh
pnpm add @kitsuyui/react-wavebox
```

## License

MIT
