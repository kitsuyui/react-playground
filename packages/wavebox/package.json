{
  "name": "@kitsuyui/react-wavebox",
  "version": "0.0.0",
  "description": "A component container that shape is changing time to time.",
  "license": "MIT",
  "author": "Yui Kitsu <kitsuyui@kitsuyui.com>",
  "exports": {
    ".": {
      "require": {
        "type": "./dist/index.d.mts",
        "default": "./dist/index.cjs"
      },
      "import": {
        "type": "./dist/index.d.mts",
        "default": "./dist/index.mjs"
      }
    }
  },
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "package.json"
  ],
  "scripts": {
    "build": "tsup --config ../../tsup.config.mjs --clean",
    "dev": "tsup --config ../../tsup.config.mjs --watch"
  },
  "devDependencies": {
    "@types/node": "^22.15.20",
    "@types/react": "^19.1.4",
    "@types/react-dom": "^19.1.5",
    "react": "^19.1.0",
    "react-dom": "^19.1.0",
    "react-use": "^17.6.0"
  },
  "peerDependencies": {
    "react": "*",
    "react-dom": "*",
    "react-use": "*"
  }
}
