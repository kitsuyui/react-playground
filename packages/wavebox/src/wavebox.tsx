import React from 'react'
import { useInterval } from 'react-use'
import { curve } from './curve'
import { SizedBox } from './SizedBox'

const DEFAULT_INTERVAL = 10  // Default interval in milliseconds for the wave animation (10ms)

/**
 * WaveBox component that animates its width and height in a wave-like manner.
 *
 * @param props - The properties for the WaveBox component.
 * @param props.children - The content to be rendered inside the WaveBox.
 * @param props.maxWidth - The maximum width of the WaveBox.
 * @param props.maxHeight - The maximum height of the WaveBox.
 * @param props.minWidth - The minimum width of the WaveBox.
 * @param props.minHeight - The minimum height of the WaveBox.
 * @param props.widthCycle - The cycle duration for the width animation.
 * @param props.heightCycle - The cycle duration for the height animation.
 * @param props.interval - The interval in milliseconds for updating the wave animation (default is 10ms).
 * @returns {React.JSX.Element} The rendered WaveBox component.
 */
export const WaveBox = (props: {
  children: React.ReactNode
  maxWidth: number
  maxHeight: number
  minWidth: number
  minHeight: number
  widthCycle: number
  heightCycle: number
  interval?: number
}): React.JSX.Element => {
  const { children, maxWidth, maxHeight, minWidth, minHeight, widthCycle, heightCycle } = props
  const interval = props.interval ?? DEFAULT_INTERVAL
  return (
    <WaveBoxProvider
      maxWidth={maxWidth}
      maxHeight={maxHeight}
      minWidth={minWidth}
      minHeight={minHeight}
      widthCycle={widthCycle}
      heightCycle={heightCycle}
      interval={interval}
    >
      <BaseWaveBox>{children}</BaseWaveBox>
    </WaveBoxProvider>
  )
}

/**
 * WaveBoxContext that provides the context for the WaveBox.
 * It contains the properties needed to calculate the wave-like behavior.
 */
export const WaveBoxContext = React.createContext<{
  maxWidth: number
  maxHeight: number
  minWidth: number
  minHeight: number
  widthCycle: number
  heightCycle: number
  tick: number
}>({
  maxWidth: 0,
  maxHeight: 0,
  minWidth: 0,
  minHeight: 0,
  widthCycle: 0,
  heightCycle: 0,
  tick: 0,
})

/**
 * WaveBoxProvider component that provides the context for the WaveBox.
 * It manages the tick state that is used to animate the wave-like behavior.
 * @param props.children - The content to be rendered inside the WaveBoxProvider.
 * @param props.maxWidth - The maximum width of the WaveBox.
 * @param props.maxHeight - The maximum height of the WaveBox.
 * @param props.minWidth - The minimum width of the WaveBox.
 * @param props.minHeight - The minimum height of the WaveBox.
 * @param props.widthCycle - The cycle duration for the width animation.
 * @param props.heightCycle - The cycle duration for the height animation.
 * @param props.interval - The interval in milliseconds for updating the tick state (default is 10ms).
 * @returns {React.JSX.Element} The rendered WaveBoxProvider component.
 */
export const WaveBoxProvider = (props: {
  children: React.ReactNode
  maxWidth: number
  maxHeight: number
  minWidth: number
  minHeight: number
  widthCycle: number
  heightCycle: number
  interval?: number
}): React.JSX.Element => {
  const { children, maxWidth, maxHeight, minWidth, minHeight, widthCycle, heightCycle } = props
  const interval = props.interval ?? DEFAULT_INTERVAL
  const [tick, setTick] = React.useState<number>(0)
  useInterval(() => {
    setTick((prevTick) => prevTick + 1)
  }, interval)
  return (
    <WaveBoxContext
      value={{ maxWidth, maxHeight, minWidth, minHeight, widthCycle, heightCycle, tick }}
    >
      {children}
    </WaveBoxContext>
  )
}

/**
 * BaseWaveBox component that uses the WaveBoxContext to determine its size based on the sine wave curve.
 * @param props - The properties for the BaseWaveBox component.
 * @returns {React.JSX.Element} The rendered BaseWaveBox component.
 */
const BaseWaveBox = (props: {
  children: React.ReactNode
}): React.JSX.Element => {
  const { maxWidth, maxHeight, minWidth, minHeight, widthCycle, heightCycle, tick } = React.useContext(WaveBoxContext)
  const currentWidth = curve(minWidth, maxWidth, widthCycle, tick)
  const currentHeight = curve(minHeight, maxHeight, heightCycle, tick)
  return (
    <SizedBox width={currentWidth} height={currentHeight}>
      {props.children}
    </SizedBox>
  )
}
