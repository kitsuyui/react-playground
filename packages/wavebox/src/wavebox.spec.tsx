import { expect, describe, it } from 'vitest'

import { render, cleanup } from '@testing-library/react'

import { WaveBox, WaveBoxContext, WaveBoxProvider } from './wavebox'
import React from 'react'

describe('WaveBox', () => {
  it('renders with specified width and height', () => {
    const { asFragment } = render(
      <WaveBox
        minWidth={100}
        maxWidth={200}
        minHeight={100}
        maxHeight={200}
        widthCycle={100}
        heightCycle={100}
      >
        <div>Test Content</div>
      </WaveBox>
    )
    expect(asFragment()).toMatchSnapshot()
    // Avoid useInterval cleanup issues in tests
    cleanup()
  })
  it('accepts custom interval', () => {
    const { asFragment } = render(
      <WaveBox
        minWidth={100}
        maxWidth={200}
        minHeight={100}
        maxHeight={200}
        widthCycle={100}
        heightCycle={100}
        interval={50} // Custom interval
      >
        <div>Test Content</div>
      </WaveBox>
    )
    expect(asFragment()).toMatchSnapshot()
    // Avoid useInterval cleanup issues in tests
    cleanup()
  })
})

describe('WaveBoxContext', () => {
  it('provides context values', () => {
    const WaveBoxContextProviderTester = () => {
      const context = React.useContext(WaveBoxContext)
      return (
        <div>
          {JSON.stringify({
            maxWidth: context.maxWidth,
            maxHeight: context.maxHeight,
            minWidth: context.minWidth,
            minHeight: context.minHeight,
            widthCycle: context.widthCycle,
            heightCycle: context.heightCycle,
            tick: context.tick,
          }, null, 2)}
        </div>
      )
    }
    const { asFragment } = render(
      <WaveBoxProvider
        maxWidth={200}
        maxHeight={200}
        minWidth={100}
        minHeight={100}
        widthCycle={100}
        heightCycle={100}
      >
        <WaveBoxContextProviderTester />
      </WaveBoxProvider>
    )
    expect(asFragment()).toMatchSnapshot()
    // Avoid useInterval cleanup issues in tests
    cleanup()
  })
})
