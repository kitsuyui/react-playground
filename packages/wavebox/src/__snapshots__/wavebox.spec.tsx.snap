// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WaveBox > accepts custom interval 1`] = `
<DocumentFragment>
  <div
    style="width: 150px; height: 150px; overflow: hidden; position: relative;"
  >
    <div>
      Test Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`WaveBox > renders with specified width and height 1`] = `
<DocumentFragment>
  <div
    style="width: 150px; height: 150px; overflow: hidden; position: relative;"
  >
    <div>
      Test Content
    </div>
  </div>
</DocumentFragment>
`;

exports[`WaveBoxContext > provides context values 1`] = `
<DocumentFragment>
  <div>
    {
  "maxWidth": 200,
  "maxHeight": 200,
  "minWidth": 100,
  "minHeight": 100,
  "widthCycle": 100,
  "heightCycle": 100,
  "tick": 0
}
  </div>
</DocumentFragment>
`;
